lib/irb/ruby-lex.rb

```diff
@@ -214,67 +213,50 @@ def check_state(code, tokens)
     [ltype, indent, continue, code_block_open]
   end
 
-  def prompt
-    if @prompt
-      @prompt.call(@ltype, @indent, @continue, @line_no)
-    end
+  def save_prompt_to_context_io(ltype, indent, continue, line_num_offset)
+    # Implicitly saves prompt string to `@context.io.prompt`. This will be used in the next `@input.call`.
+    @prompt.call(ltype, indent, continue, @line_no + line_num_offset)
```

Comment:
`@prompt` is not going to be `nil` in actual use cases because `Irb#eval_input` always calls `RubyLex#set_prompt` before calling `RubyLex#each_top_level_statemen`.
But I think this type of implicit dependency is risky and we should probably refactor that later.